src/ILVerification/src/TypeVerifier.cs

```diff
@@ -98,36 +99,76 @@ public void VerifyInterfaces()
                         MethodDesc resolvedMethod = type.ResolveInterfaceMethodTarget(method);
                         if (resolvedMethod is null)
                         {
-                            VerificationError(VerifierError.InterfaceMethodNotImplemented, Format(type), Format(implementedInterface.DefType), Format(method));
+                            VerificationError(VerifierError.InterfaceMethodNotImplemented, Format(type), Format(implementedInterface.DefType, implementedInterface.InterfaceImplementation), Format(method));
```

Comment:
I would avoid mixing module/token pairs coming from TypeDescs with module/token pairs coming from metadata. Either:

- Get both the module and token from TypeDesc
     or
- Get module matching the metadata reader (`_module` in this case), and use the original metadata handle to get the token. No TypeDescs involved.

Reply:
> Get module matching the metadata reader (_module in this case), and use the original metadata handle to get the token. No TypeDescs involved.

I did it:
```
[MD]: Error: Class implements interface but not method, Class: '[InterfaceImplementation]InterfaceImplHasDuplicateAndMissingMethod_InvalidType_InterfaceImplHasDuplicate@InterfaceMethodNotImplemented([InterfaceImplementation]0x02000006)' Interface: '[InterfaceDefinition]Interface([InterfaceImplementation]0x09000006)' Method: '[InterfaceDefinition]Interface.M1()([InterfaceImplementation]0x06000001)'.
```
([InterfaceImplementation]0x02000006) -> metadata reader module, typedef token 
([InterfaceDefinition]Interface([InterfaceImplementation]0x09000006) -> metadata reader module, interface implementation token
([InterfaceImplementation]0x06000001) -> metadata reader module, method token

generics sample:
[GenericInterface]GenericsMissingMethod_InvalidType_InterfaceMethodNotImplemented([GenericInterface]0x02000004) -> metadata reader module, typedef token
[GenericInterface]IGeneric`2<int32,int32>([GenericInterface]0x09000002) -> metadata reader module, interface implementation token
[GenericInterface]IGeneric`2<int32,int32>.M1(int32,int32)([GenericInterface]0x06000001) metadata reader module, method token

Am I close to your idea?

Reply:
The tokens seems to be still mixed up.

You should be able to check the tokens using `ildasm /tokens`. When you run ildasm on the token module and find that token in the ildasm output, you should get the item that is human readable description refers to.

Let's take `Method: '[InterfaceDefinition]Interface.M1()([InterfaceImplementation]0x06000001)'`.

If you run `ildasm /tokens InterfaceImplementation.dll /out:x.txt` and find method with token `0x06000001`, you should get the method `M1()`. It won't be the case if I am reading the code correctly.

Reply:
Never give up!Thank's for tips and patience!I'm pretty newbies to raw ecma metadata world and sometimes is not easy to me to relate book with class code!